File: apps/web/src/matrix/specFeatures.test.ts
import { describe, expect, it } from "vitest";
import {
  attachMentions,
  belongsOnMainTimeline,
  completeCommand,
  conversationReplyContent,
  filterCommandSuggestions,
  intentionalMentions,
  locationContent,
  openStreetMapUrl,
  parseCommandsState,
  parseGeoUri,
  parseImagePack,
  parseLocationContent,
  parseMsc4139Prompts,
  parseRoomSummary,
  slidingSyncSupported,
  stickerContent,
  threadRelation,
  threadRootId,
  threadSubscriptionPath,
} from "./specFeatures";

describe("intentional mentions (MSC3952)", () => {
  it("collects mentioned mxids and @room", () => {
    expect(
      intentionalMentions("hey @alice:example.org and @room", [
        "@alice:example.org",
        "@bob:example.org",
      ]),
    ).toEqual({ user_ids: ["@alice:example.org"], room: true });
  });

  it("still writes m.mentions when nobody is tagged", () => {
    expect(attachMentions({ msgtype: "m.text", body: "hi" }, "hi", ["@alice:example.org"])).toEqual({
      msgtype: "m.text",
      body: "hi",
      "m.mentions": { user_ids: [] },
    });
  });
});

describe("threads (MSC3440)", () => {
  it("builds an m.thread relation with fallback for the main timeline", () => {
    expect(threadRelation("$root")).toEqual({
      rel_type: "m.thread",
      event_id: "$root",
      is_falling_back: true,
      "m.in_reply_to": { event_id: "$root" },
    });
    expect(threadRelation("$root", "$root", false).is_falling_back).toBe(false);
  });

  it("keeps in-thread replies off the main timeline unless they are fallbacks", () => {
    const inThread = {
      "m.relates_to": { rel_type: "m.thread", event_id: "$root", is_falling_back: false },
    };
    const fallback = {
      "m.relates_to": { rel_type: "m.thread", event_id: "$root", is_falling_back: true },
    };
    expect(belongsOnMainTimeline(inThread)).toBe(false);
    expect(belongsOnMainTimeline(fallback)).toBe(true);
    expect(threadRootId(inThread)).toBe("$root");
  });
});

describe("location (MSC3488)", () => {
  it("writes geo_uri plus the extensible location block", () => {
    const content = locationContent(55.75, 37.62, "Red Square");
    expect(content.msgtype).toBe("m.location");
    expect(content.geo_uri).toBe("geo:55.75,37.62");
    expect(parseLocationContent(content)).toEqual({
      lat: 55.75,
      lon: 37.62,
      geoUri: "geo:55.75,37.62",
      description: "Red Square",
    });
    expect(parseGeoUri("geo:1,2;u=30")).toEqual({ lat: 1, lon: 2 });
    expect(openStreetMapUrl(1, 2)).toContain("mlat=1");
  });
});

describe("image packs (MSC2545)", () => {
  it("reads pack images and builds m.sticker content", () => {
    const items = parseImagePack({
      images: {
        wave: { url: "mxc://example.org/abc", body: "wave", usage: ["sticker"] },
        skip: { url: "https://evil.example/x.png" },
      },
    });
    expect(items).toEqual([
      { shortcode: "wave", url: "mxc://example.org/abc", body: "wave", usage: ["sticker"] },
    ]);
    expect(stickerContent(items[0]!)).toMatchObject({ body: "wave", url: "mxc://example.org/abc" });
  });
});

describe("bot commands (MSC4332) and prompts (MSC4139)", () => {
  it("parses vendor and MSC command state", () => {
    const vendor = parseCommandsState({
      commands: [{ name: "start", aliases: ["go"], description: "Begin" }],
    });
    const msc = parseCommandsState({
      "org.matrix.msc4332.commands": {
        commands: [{ name: "roll", aliases: ["dice"] }],
      },
    });
    expect(vendor[0]?.name).toBe("start");
    expect(msc[0]?.name).toBe("roll");
    expect(filterCommandSuggestions([...vendor, ...msc], "/r").map((c) => c.name)).toEqual(["roll"]);
    expect(completeCommand(vendor[0]!, "!st")).toBe("!start ");
  });

  it("parses MSC4139 prompts and builds a conversation reply", () => {
    const parsed = parseMsc4139Prompts({
      "org.matrix.msc4139.prompts": {
        intro: { content: { "m.text": [{ body: "Pick a die" }] } },
        scope: ["@alice:example.org"],
        prompts: [
          { type: "preset", id: "1d6", label: { "m.text": [{ body: "1d6" }] } },
          { type: "input", id: "custom", validator: "[0-9]+d[0-9]+", label: { "m.text": [{ body: "Other" }] } },
        ],
      },
    });
    expect(parsed?.prompts).toHaveLength(2);
    expect(parsed?.prompts[0]).toEqual({ id: "1d6", type: "preset", label: "1d6" });
    const reply = conversationReplyContent("1d6", "1d6", "$welcome");
    expect(reply["org.matrix.msc4139.used_prompt"]).toEqual({ id: "1d6" });
    expect((reply["m.relates_to"] as { rel_type: string }).rel_type).toBe("m.thread");
  });
});

describe("homeserver helpers", () => {
  it("detects simplified sliding sync flags", () => {
    expect(slidingSyncSupported({ "org.matrix.simplified_msc3575": true })).toBe(true);
    expect(slidingSyncSupported({ "org.matrix.msc4186": true })).toBe(true);
    expect(slidingSyncSupported({})).toBe(false);
  });

  it("builds the MSC4306 subscription path and room summary", () => {
    expect(threadSubscriptionPath("!r:ex", "$t")).toContain("msc4306");
    expect(parseRoomSummary({ name: "Lobby", num_joined_members: 4 }, "!x:ex")).toMatchObject({
      roomId: "!x:ex",
      name: "Lobby",
      numJoinedMembers: 4,
    });
  });
});

File: apps/client/test/spec_features_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:highlife_client/domain/spec_features.dart';

void main() {
  test('intentional mentions collect mxids and @room', () {
    final mentions = intentionalMentions('hey @alice:example.org and @room', [
      '@alice:example.org',
      '@bob:example.org',
    ]);
    expect(mentions.userIds, ['@alice:example.org']);
    expect(mentions.room, isTrue);
  });

  test('thread replies stay off the main timeline unless fallback', () {
    expect(
      belongsOnMainTimeline({
        'm.relates_to': {
          'rel_type': 'm.thread',
          'event_id': 'root',
          'is_falling_back': false,
        },
      }),
      isFalse,
    );
    expect(
      belongsOnMainTimeline({
        'm.relates_to': {
          'rel_type': 'm.thread',
          'event_id': 'root',
          'is_falling_back': true,
        },
      }),
      isTrue,
    );
    expect(threadRelation('root')['is_falling_back'], isTrue);
    expect(
      threadRelation('root', replyToId: 'root', fallback: false)['is_falling_back'],
      isFalse,
    );
  });

  test('location content round-trips MSC3488', () {
    final content = locationContent(55.75, 37.62, description: 'Red Square');
    expect(content['msgtype'], 'm.location');
    final parsed = parseLocationContent(Map<String, dynamic>.from(content));
    expect(parsed?.lat, 55.75);
    expect(parsed?.description, 'Red Square');
  });

  test('image pack parser ignores non-mxc urls', () {
    final items = parseImagePack({
      'images': {
        'wave': {'url': 'mxc://example.org/abc', 'body': 'wave'},
        'skip': {'url': 'https://evil.example/x.png'},
      },
    });
    expect(items, hasLength(1));
    expect(stickerContent(items.single)['url'], 'mxc://example.org/abc');
  });

  test('MSC4139 prompts parse preset buttons', () {
    final parsed = parseMsc4139Prompts({
      msc4139PromptsKey: {
        'prompts': [
          {
            'type': 'preset',
            'id': '1d6',
            'label': {
              'm.text': [
                {'body': '1d6'},
              ],
            },
          },
        ],
      },
    });
    expect(parsed?.prompts.single.id, '1d6');
    expect(conversationReplyContent(promptId: '1d6', label: '1d6', rootEventId: r'$w')['body'], '1d6');
  });

  test('sliding sync flags and subscription path', () {
    expect(slidingSyncSupported({'org.matrix.simplified_msc3575': true}), isTrue);
    expect(threadSubscriptionPath('!r:ex', r'$t'), contains('msc4306'));
    expect(parseRoomSummary({'name': 'Lobby'}, '!x:ex').name, 'Lobby');
    expect(matrixApiAction(threadSubscriptionPath('!r:ex', r'$t')), startsWith('/client/'));
    expect(
      parseSlidingSyncRoomOrder({
        'lists': {
          'all': {
            'ops': [
              {
                'op': 'SYNC',
                'room_ids': ['!a:ex', '!b:ex'],
              },
            ],
          },
        },
      }),
      ['!a:ex', '!b:ex'],
    );
    expect(parseProfileAbout({profileAboutKey: 'hello from HighLife'}), 'hello from HighLife');
    expect(firstHttpUrl('see https://example.org/a, then text'), 'https://example.org/a');
    expect(
      parseUrlPreview({'og:title': 'Example'}, 'https://example.org')?.title,
      'Example',
    );
  });

  test('MSC4332 commands parse nested and top-level command lists', () {
    final nested = parseCommandsState({
      msc4332CommandsState: {
        'commands': [
          {'name': 'roll', 'aliases': ['r'], 'description': 'dice'},
        ],
      },
    });
    expect(nested.single.name, 'roll');
    expect(nested.single.aliases, ['r']);
    final top = parseCommandsState({
      'commands': [
        {'name': 'help'},
      ],
    });
    expect(top.single.name, 'help');
  });

  test('login QR payload stays a local matrix: URI', () {
    expect(
      matrixLoginQrPayload(
        homeserver: 'https://matrix.example.org',
        deviceId: 'DEVICE',
        userId: '@alice:example.org',
      ),
      startsWith('matrix://login?'),
    );
  });
}
